toolchain/lower/coalesce.h

```diff
@@ -0,0 +1,155 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_LOWER_COALESCE_H_
+#define CARBON_TOOLCHAIN_LOWER_COALESCE_H_
+
+#include "llvm/Support/BLAKE3.h"
+#include "toolchain/lower/context.h"
+#include "toolchain/sem_ir/ids.h"
+
+namespace Carbon::Lower {
+
+// Coalescing functionality for lowering fewer specifics of the same generic.
+class CoalesceSpecifics {
+ public:
+  using LoweredSpecificsStore =
+      FixedSizeValueStore<SemIR::GenericId,
+                          llvm::SmallVector<SemIR::SpecificId>>;
+  using LoweredLlvmFunctionStore =
+      FixedSizeValueStore<SemIR::SpecificId, llvm::Function*>;
+
+  // Describes a specific function's body fingerprint.
+  struct SpecificFunctionFingerprint {
+    // Fingerprint with all specific-dependent instructions, except specific
+    // calls. This is built by the `FunctionContext` while lowering each
+    // instruction in the definition of a specific function.
+    // TODO: This can be merged with the function type fingerprint, for a
+    // single upfront non-equivalence check, and hash bucketing for deeper
+    // equivalence evaluation.
+    llvm::BLAKE3Result<32> common_fingerprint;
+    // Fingerprint for all calls to specific functions (hashes all calls to
+    // other specifics). This is built by the `FunctionContext` while lowering.
+    llvm::BLAKE3Result<32> specific_fingerprint;
+    // All non-hashed specific_ids of functions called.
+    llvm::SmallVector<SemIR::SpecificId> calls;
+  };
+
+  explicit CoalesceSpecifics(llvm::raw_ostream* vlog_stream,
```

Comment:
Maybe document what's done with `specifics` (I was trying to figure out if it was stored):

```suggestion
  // Takes a `SpecificStore` to help initialize related `FixedSizeValueStore`s.
  explicit CoalesceSpecifics(llvm::raw_ostream* vlog_stream,
```

Note, could also just pass in the size, but I think it's fair enough to pass in the type to get the related bits of safety.